[GLib] Add API tests for WebKitXRPermissionRequest

fujii opened the pull request:
#### cfcefc3019b7a9e6e2e643a2d6cf6529debf69be
<pre>
[GLib] Add API tests for WebKitXRPermissionRequest
<a href="https://bugs.webkit.org/show_bug.cgi?id=299585">https://bugs.webkit.org/show_bug.cgi?id=299585</a>

Reviewed by Adrian Perez de Castro.

Added API tests for WebKitXRPermissionRequest.

featureRequiresExplicitConsent() in
Source/WebCore/Modules/webxr/WebXRSystem.cpp returns true only for
`PlatformXR::SessionFeature::HandTracking`. Thus, hand-tracking is the
only explicit consent required feature. All other features are always
implicitly granted.

However, Monado doesn&apos;t support hand-tracking. We can&apos;t make a
meaningful API test for consent required features and consent optional
features at the moment.

* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp:
(testWebKitXRPermissionRequest):
(beforeAll):
* Tools/TestWebKitAPI/glib/TestExpectations.json:

Canonical link: <a href="https://commits.webkit.org/300621@main">https://commits.webkit.org/300621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c227528459b9d9e5993e2c002cda8c0c4a9747da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123273 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42987 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33684 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129981 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43711 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51582 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93718 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126226 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34827 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110294 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33791 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28452 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73499 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104541 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132699 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50223 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38215 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102191 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50599 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106517 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102046 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47400 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25617 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46997 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19420 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50078 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52899 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp:
```diff
@@ -130,12 +131,116 @@ static void testWebKitWebXRLeaveImmersiveModeAndWaitUntilImmersiveModeChanged(We
+    g_signal_connect(test->webView(), "permission-request", G_CALLBACK(permissionRequestCallback), &data);
+
+    WebViewTest::NetworkPolicyGuard guard(test, WEBKIT_TLS_ERRORS_POLICY_IGNORE);
+    test->loadHtml("", "https://foo.com/bar");
```
This is much cleaner and simpler than what I did in the other test, nice

spenap commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp:
```diff
@@ -130,12 +131,116 @@ static void testWebKitWebXRLeaveImmersiveModeAndWaitUntilImmersiveModeChanged(We
+        data.answer = answer;
+        data.resetResult();
+        test->runJavaScriptAndWaitUntilFinished(script.utf8().data(), nullptr);
+        test->waitUntilTitleChanged();
```
Nice, too. I can follow this approach by updating the title if the session fails to start and raises an error -- which it shouldn't unless the environment is not well setup.

spenap reviewed (commented):
This is a very nice test, and I'm looking forward to updating the other one with a similar approach to what you took

fujii committed:
[GLib] Add API tests for WebKitXRPermissionRequest
https://bugs.webkit.org/show_bug.cgi?id=299585

Reviewed by Adrian Perez de Castro.

Added API tests for WebKitXRPermissionRequest.

featureRequiresExplicitConsent() in
Source/WebCore/Modules/webxr/WebXRSystem.cpp returns true only for
`PlatformXR::SessionFeature::HandTracking`. Thus, hand-tracking is the
only explicit consent required feature. All other features are always
implicitly granted.

However, Monado doesn't support hand-tracking. We can't make a
meaningful API test for consent required features and consent optional
features at the moment.

* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebXR.cpp:
(testWebKitXRPermissionRequest):
(beforeAll):
* Tools/TestWebKitAPI/glib/TestExpectations.json:

Canonical link: https://commits.webkit.org/300621@main
